### PR TITLE
OrbitControls: allow for inverse Azimuthal range

### PR DESCRIPTION
**Pull request by sciecode**
**Motive**

Improve OrbitControls to allow any combination of min/max azimuth angle. So that camera doesn't need to be looking towards Z- to work correctly.

**Discussion**

In the current setup, we only allow for `minAzimuthAngle` to be less than `maxAzimuthAngle` . However this limits the way we can setup and adjust the camera view range. The following is how we currently have it setup.

However,  it doesn't allow for the inverse range:

This is because, as soon as `MaxAzimuthAngle` becomes smaller than `MinAzimuthAngle`, the range test is no longer valid. AKA, If either min or max is crosses over PI and gets inverted - to + or + to -.

In other words, if min or max don't cross PI, we want the azimuthal angle to be within range [ min, max ]. However, once either one crosses PI, we want azimuthal angle to be within range [ - Infinity, max ]  or  range[ min, infinity ].

**Final thoughts**

We usually recommend that the user rotate the scene to be facing Z-, however if the scene has many cameras it might be impossible for all of the cameras to be within a valid configuration.

This doesn't alter previous behavior ( backwards compatible ), it just adds the correct behavior for the inversed range.

### CODE REVIEW

**Comment by WestLangley** [comment]
The proposed API requires the user to set `maxAzimuthAngle` < `minAzimuthAngle`.

Instead, just support a wider interval:
```
// How far you can orbit horizontally, upper and lower limits.
// If set, must be a sub-interval of the interval [ - 2 * Math.PI, 2 * Math.PI ].
this.minAzimuthAngle = - Infinity; // radians
this.maxAzimuthAngle = Infinity; // radians
```

**Comment by sciecode** [comment]
>  just support a wider interval

That would probably be the best API for this, but I'm afraid that's not possible without extending `Spherical.theta` to allow the same range. It also doesn't make sense in the context of `Spherical` design.

As discussed privately, the other approach would be to change the API entirely to something like:

    this.azimuthCenter = 0; // range center angle [ - PI, PI ]
    this.azimuthRange = 2 * Math.PI; // total angulation range [ 0, TWO_PI ]

The implementation is not backwards compatible and may conflict with `controls.min|maxPolarAngle` API. The latter, because it does not make sense to support such an API when we can't cross the poles.

However, it allows us to automate the initial angle based on target orientation, users will mostly just need to care about `azimuthRange`. Which, I think, is kind of a nice default behavior.

**Comment by Mugen87** [comment]
> However, it allows us to automate the center angle based on target orientation, users will mostly just need to care about azimuthRange. Which, I think, is kind of a nice default behavior.

I don't think it's worth to break existing code for your proposed feature. It would be great if you can find a backwards compatibility implementation.

**Comment by sciecode** [comment]
> It would be great if you can find a backwards compatibility implementation. 

I did find a backwards compatible implementation. It is this PR.

**Comment by Mugen87** [comment]
Sorry, I should have written my post more clear. I just wanted to vote a against a breaking change. However @WestLangley concerns should be addressed, too.

**Comment by munrocket** [comment]
Took ideas from @sciecode , @WestLangley and continued PR in #19296.
After some war with CI I get result with `isFenite()` function that seems right.

**Comment by Mugen87** [comment]
Do we still need this PR with regard to #19298? It would be great if it possible to agree on a single implementation (if this feature is actually wanted).

**Comment by sciecode** [comment]
As I said, on https://github.com/mrdoob/three.js/pull/19298#issuecomment-623743394 I'm fine with either one, I just dislike the API. I would love to have something like `setAzimuthAngle` or something that can be calculated at invocation time, rather than having to run multiple checks on each frame update.

We can go with WestLangley's approach, it doesn't involve reversing the limits, whereas mine does.

**Comment by Mugen87** [comment]
> We can go with WestLangley's approach, it doesn't involve reversing the limits, whereas mine does.

Okay, closing then.
